### PR TITLE
Upgrade pants from 2.14 to 2.15

### PR DESCRIPTION
**Pull request by cognifloyd**
### Background

This is another part of introducing [pants](https://www.pantsbuild.org/docs), as discussed in various TSC meetings.

Related PRs can be found in:
- [pantsbuild](https://github.com/StackStorm/st2/milestone/47) milestone (which includes PRs since #5713)
- https://github.com/StackStorm/st2/labels/pantsbuild label (which also includes preparatory PRs that came before adding pantsbuild)

### Overview of this PR

This updates to pants to `2.15.0` which requires:
- regen the pants-plugins lockfile
- several changes to our `fmt` and `lint` plugins in pants-plugins.

The required plugin updates are described in the pants [Plugin upgrade guide](https://www.pantsbuild.org/v2.15/docs/plugin-upgrade-guide).

#### Notes about updating to pants 2.15.0

Here are some links to learn more about pants v2.15:
- https://blog.pantsbuild.org/pants-2-15/
- https://github.com/pantsbuild/pants/blob/main/src/python/pants/notes/2.15.x.md

NB: We must do updates 1 minor version of pants at a time so that pants `update-build-files` goal can handle any required updates before continuing. We want to go to 2.16, but the 2.15 update was significant enough that it deserves its own PR before jumping to 2.16.

_Aside: As soon as the `pants 2.16.0rc0` comes out, I want to switch to that one as we need some of the features in 2.16 like the new `stevedore` backend, and improvements to exporting virtualenvs based on lockfiles. That does not look like it will require extensive plugin upgrades like the 2.15 upgrade does._

### The Fmt/Lint plugin API changed in pants 2.15

It now requires a Subsystem with a SkipOption. I took advantage of this to consolidate the constants and all the address and pex request definitions in the Subsystem subclass.

It now requires 2 rules instead of 1: to partition, then fmt/lint. That means the structure of the rule requests had to change slightly. The tests also had to account for the modified rules.

_Refer to the [Plugin upgrade guide](https://www.pantsbuild.org/v2.15/docs/plugin-upgrade-guide) for more details._

### CODE REVIEW

**Comment by cognifloyd** [comment]
Aargh. Something isn't working in my update. Tests pass, but actually running the linters gets caught in some weird loop.

File: pants-plugins/api_spec/rules.py [hunk 1]
```diff
@@ -63,79 +50,39 @@ class APISpecFieldSet(FieldSet):
...
     desc="Update openapi.yaml with st2-generate-api-spec",
     level=LogLevel.DEBUG,
 )
 async def generate_api_spec_via_fmt(
-    request: GenerateAPISpecViaFmtTargetsRequest,
+    request: GenerateAPISpecViaFmtTargetsRequest.Batch,
+    subsystem: GenerateApiSpec,
 ) -> FmtResult:
     # There will only be one target+field_set, but we iterate
     # to satisfy how fmt expects that there could be more than one.
```
Comment (amanda11) [inline pants-plugins/api_spec/rules.py:72]: Does this comment make sense now we have removed this chunk of code?

File: pants-plugins/api_spec/rules.py [hunk 2]
```diff
@@ -63,79 +50,39 @@ class APISpecFieldSet(FieldSet):
...
-    )
-
-    source_files_get = Get(
-        SourceFiles,
-        SourceFilesRequest(field_set.source for field_set in request.field_sets),
-    )
+    config_files_get = Get(ConfigFiles, ConfigFilesRequest, subsystem.config_request())
 
     # actually generate it with an external script.
     # Generation cannot be inlined here because it needs to import the st2 code.
```
Comment (amanda11) [inline pants-plugins/api_spec/rules.py:78]: Does this comment make sense now we have removed this chunk of code, and so we no-longer directly call GENERATE_CMD python script here?

File: pants-plugins/api_spec/rules.py [hunk 3]
```diff
@@ -144,87 +91,59 @@ async def generate_api_spec_via_fmt(
...
     desc="Validate openapi.yaml with st2-validate-api-spec",
     level=LogLevel.DEBUG,
 )
 async def validate_api_spec(
-    request: ValidateAPISpecRequest,
-) -> LintResults:
+    request: ValidateAPISpecRequest.Batch,
+    subsystem: ValidateApiSpec,
+) -> LintResult:
     # There will only be one target+field_set, but we iterate
```
Comment (amanda11) [inline pants-plugins/api_spec/rules.py:125]: Does this comment make sense now we have removed this chunk of code?

File: pants-plugins/api_spec/rules.py [hunk 4]
```diff
@@ -144,87 +91,59 @@ async def generate_api_spec_via_fmt(
...
-
     source_files_get = Get(
         SourceFiles,
-        SourceFilesRequest(field_set.source for field_set in request.field_sets),
+        SourceFilesRequest(field_set.source for field_set in request.elements),
     )
 
+    config_files_get = Get(ConfigFiles, ConfigFilesRequest, subsystem.config_request())
+
     # actually validate it with an external script.
```
Comment (amanda11) [inline pants-plugins/api_spec/rules.py:136]: Does this comment make sense now we have removed this chunk of code, as we nolonger have code here to call VALIDATE_CMD python module

File: pants-plugins/sample_conf/rules.py [hunk 5]
```diff
@@ -51,37 +46,25 @@ class GenerateSampleConfFieldSet(FieldSet):
...
-    request: GenerateSampleConfViaFmtTargetsRequest,
+    request: GenerateSampleConfViaFmtTargetsRequest.Batch,
+    subsystem: ConfigGen,
 ) -> FmtResult:
     # There will only be one target+field_set, but we iterate
     # to satisfy how fmt expects that there could be more than one.
     # If there is more than one, they will all get the same contents.
 
     # actually generate it with an external script.
     # Generation cannot be inlined here because it needs to import the st2 code.
```
Comment (amanda11) [inline pants-plugins/sample_conf/rules.py:66]: Does this comment make sense now we have removed this chunk of code?

File: pants-plugins/schemas/rules.py [hunk 6]
```diff
@@ -49,37 +42,24 @@ class GenerateSchemasFieldSet(FieldSet):
...
     desc="Update contrib/schemas/*.json with st2-generate-schemas",
     level=LogLevel.DEBUG,
 )
 async def generate_schemas_via_fmt(
-    request: GenerateSchemasViaFmtTargetsRequest,
+    request: GenerateSchemasViaFmtTargetsRequest.Batch,
+    subsystem: GenerateSchemas,
 ) -> FmtResult:
     # We use a pex to actually generate the schemas with an external script.
     # Generation cannot be inlined here because it needs to import the st2 code.
```
Comment (amanda11) [inline pants-plugins/schemas/rules.py:58]: Does this comment make sense now we have removed this chunk of code, as the external script isnt' directly referenced here anymore?

File: pants-plugins/api_spec/rules.py (hunk 1 above)
Comment (cognifloyd) [reply pants-plugins/api_spec/rules.py]: Hmm. Yup. This is out-of-date. I'll drop the comment.

File: pants-plugins/api_spec/rules.py (hunk 3 above)
Comment (cognifloyd) [reply pants-plugins/api_spec/rules.py]: Hmm. Yup. This is out-of-date. I'll drop the comment.

File: pants-plugins/sample_conf/rules.py (hunk 5 above)
Comment (cognifloyd) [reply pants-plugins/sample_conf/rules.py:62]: The "only one target..." bit is out of date. I dropped it.

This code block was moved, not removed. The PexFromTargetsRequest (including all of the random constants used to build it) are now defined on the subsystem. I added a comment to help clarify that.

File: pants-plugins/schemas/rules.py (hunk 6 above)
Comment (cognifloyd) [reply pants-plugins/schemas/rules.py:58]: This code block was moved, not removed. The PexFromTargetsRequest (including all of the random constants used to build it) are now defined on the subsystem. I added a comment to help clarify that.

File: pants-plugins/api_spec/rules.py (hunk 4 above)
Comment (cognifloyd) [reply pants-plugins/api_spec/rules.py]: This code block was moved, not removed. The PexFromTargetsRequest (including all of the random constants used to build it) are now defined on the subsystem. I added a comment to help clarify that.

File: pants-plugins/api_spec/rules.py (hunk 2 above)
Comment (cognifloyd) [reply pants-plugins/api_spec/rules.py:74]: This code block was moved, not removed. The PexFromTargetsRequest (including all of the random constants used to build it) are now defined on the subsystem. I added a comment to help clarify that.

**Review by amanda11 — APPROVED** [review]
Thanks for updating the comments

**Pull request merged**